bftcosi/bftcosi.go

```diff
@@ -1,798 +1,239 @@
-		bft.signRefusal = true
-		log.Error("not enough commit commitment messages")
-	}
-
-	commitment := bft.commit.Commit(bft.Suite().RandomStream(), bft.tempCommitCommit)
-	if bft.IsRoot() {
-		// do nothing:
-		// stop the processing of the round, wait the end of
-		// the "prepare" round: calls startChallengeCommit
-		return nil
+func (bft *ProtocolBFTCoSi) initCosiProtocol(phase phase) (*protocol.CoSiRootNode, error) {
+	var name string
+	if phase == phasePrep {
+		name = bft.prepCosiProtoName
+	} else if phase == phaseCommit {
+		name = bft.commitCosiProtoName
+	} else {
+		return nil, fmt.Errorf("invalid phase %v", phase)
 	}
-	return bft.SendToParent(&Commitment{
-		TYPE:       RoundCommit,
-		Commitment: commitment,
-	})
-}
 
-// handleChallengePrepare collects the challenge-messages
-func (bft *ProtocolBFTCoSi) handleChallengePrepare(msg challengePrepareChan) error {
-	if bft.isClosing() {
-		return nil
-	}
-	ch := msg.ChallengePrepare
-	if !bft.IsRoot() {
-		bft.Msg = ch.Msg
-		bft.Data = ch.Data
-		// start the verification of the message
-		// acknowledge the challenge and send it down
-		bft.prepare.Challenge(ch.Challenge)
-	}
-	go func() {
-		bft.verifyChan <- bft.VerificationFunction(bft.Msg, bft.Data)
-	}()
-	if bft.IsLeaf() {
-		return bft.startResponse(RoundPrepare)
+	pi, err := bft.CreateProtocol(name, bft.Tree())
+	if err != nil {
+		return nil, err
 	}
-	return bft.sendToChildren(&ch)
-}
+	cosiProto := pi.(*protocol.CoSiRootNode)
+	cosiProto.CreateProtocol = bft.CreateProtocol
+	// NOTE We set it to n / 10 to have every sub-leader manage 10 nodes.
+	// This setting is bad if there are thousands of nodes as the root
+	// would need to manage hundres of sub-leaders.
+	cosiProto.NSubtrees = len(bft.List()) / 10
+	cosiProto.Msg = bft.Msg
+	cosiProto.Data = bft.Data
+	cosiProto.Timeout = bft.Timeout
+
+	return cosiProto, nil
+}
+
+// Dispatch is the main logic of the BFTCoSi protocol. It runs two CoSi
+// protocols as the prepare and the commit phase of PBFT. Concretely, it does:
+// 1, wait for the prepare phase to finish
+// 2, check the signature
+// 3, if it is, start the commit phase,
+//    otherwise send an empty signature
+// 4, wait for the commit phase to finish
+// 5, send the final signature
+func (bft *ProtocolBFTCoSi) Dispatch() error {
 
-// handleChallengeCommit verifies the signature and checks if not more than
-// the threshold of participants refused to sign
-func (bft *ProtocolBFTCoSi) handleChallengeCommit(msg challengeCommitChan) error {
-	if bft.isClosing() {
-		return nil
-	}
-	ch := msg.ChallengeCommit
 	if !bft.IsRoot() {
-		bft.commit.Challenge(ch.Challenge)
+		return fmt.Errorf("non-root should not start this protocol")
 	}
 
-	// verify if the signature is correct
-	data := sha512.Sum512(ch.Signature.Msg)
-	bftPrepareSig := &BFTSignature{
-		Sig:        ch.Signature.Sig,
-		Msg:        data[:],
-		Exceptions: ch.Signature.Exceptions,
+	// prepare phase (part 2)
+	prepSig := <-bft.prepSigChan
```

Comment:
I see a lot of blocking channel reads here. Are timeouts planned for a future iteration?

If a timeout was implemented by close(bft.prepSigChan), then prepSig would end up nil here, which is not checked, but I guess/hope cost.Verify would Do The Right Thing.

Reply:
It's possible, but tricky. The bft.Timeout is not a global timeout for the whole protocol, but passed down to cosi (I will make this clear in the comments). The way that cosi uses the timeout is that if a subleader does not respond after the timeout while waiting for commits, then it'll re-arrange the tree and try a new sub-leader from the original set of children. This process is repeated until it has tried all children. It is used again while waiting for responses. So it's difficult to put a precise bound on it. I am adding a rather long timeout, which is bft.Timeout * NSubtrees, which is better than nothing.